test/extended/router/externalCertificate.go

```diff
@@ -0,0 +1,682 @@
+				route := createRouteWithEC(oc.Namespace(), "route", secret.Name, helloPodSvc, host, routev1.TLSTerminationEdge)
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), route, metav1.CreateOptions{})
+				o.Expect(err).To(o.HaveOccurred())
+				o.Expect(err.Error()).To(o.ContainSubstring(`Not found: "secrets \"my-tls-secret\" not found"`))
+			})
+
+			g.It("should not support external certificate if the secret is not of type kubernetes.io/tls", func() {
+				g.By("Creating a secret with the WRONG type (Opaque)")
+				secret, _, err := createSecret(oc.Namespace(), "my-tls-secret", corev1.SecretTypeOpaque, host) // Incorrect type
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				g.By("Creating a route")
+				route := createRouteWithEC(oc.Namespace(), "route", secret.Name, helloPodSvc, host, routev1.TLSTerminationEdge)
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), route, metav1.CreateOptions{})
+				o.Expect(err).To(o.HaveOccurred())
+				o.Expect(err.Error()).To(o.ContainSubstring(`Invalid value: "my-tls-secret": secret of type "kubernetes.io/tls" required`))
+			})
+
+			g.It("should not support external certificate if the route termination type is Passthrough", func() {
+				g.By("Creating a secret with crt/key")
+				secret, _, err := createSecret(oc.Namespace(), "my-tls-secret", corev1.SecretTypeTLS, host)
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				g.By("Creating a route with Passthrough termination")
+				passthroughRoute := createRouteWithEC(oc.Namespace(), "passthrough-route", secret.Name, helloPodSvc, host, routev1.TLSTerminationPassthrough)
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), passthroughRoute, metav1.CreateOptions{})
+				o.Expect(err).To(o.HaveOccurred())
+				o.Expect(err.Error()).To(o.ContainSubstring(`Invalid value: "my-tls-secret": passthrough termination does not support`))
+			})
+
+			g.It("should not support external certificate if inline certificate is also present", func() {
+				g.By("Creating a secret with crt/key")
+				secret, _, err := createSecret(oc.Namespace(), "my-tls-secret", corev1.SecretTypeTLS, host)
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				g.By("Creating a route")
+				route := createRouteWithEC(oc.Namespace(), "route", secret.Name, helloPodSvc, host, routev1.TLSTerminationEdge)
+				// Add inline certificate
+				route.Spec.TLS.Certificate = "my-crt"
```

Comment:
can a real certificate string be added here?

because, this might error out differently with say a validation error that "invalid certificate" and bypass the other stuff (inline + external cert being present together) we're willing to check.

Reply:
> because, this might error out differently with say a validation error that "invalid certificate" and bypass the other stuff (inline + external cert being present together) we're willing to check.

I think the certificate validation will be checked by the router, after the openshift-apiserver validation succeeds. So, technically (as per the validation logic) this should work for any string.

Reply:
> and bypass the other stuff (inline + external cert being present together) we're willing to check.

We have checks to catch this 
```go
o.Expect(err.Error()).To(o.ContainSubstring(`Invalid value: "my-tls-secret": cannot specify both tls.certificate and tls.externalCertificate`))
```